yt_dlp/extractor/niconico.py

```diff
@@ -606,8 +609,63 @@ def get_video_info_xml(items):
         }
 
 
-class NiconicoPlaylistIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?nicovideo\.jp/(?:user/\d+/|my/)?mylist/(?P<id>\d+)'
+class NiconicoPlaylistBaseIE(InfoExtractor):
+    _PAGE_SIZE = 100
+
+    _API_HEADERS = {
+        'X-Frontend-ID': '6',
+        'X-Frontend-Version': '0',
+        'X-Niconico-Language': 'en-us'
+    }
+
+    def _call_api(self, list_id, resource, query):
+        "Implement this in child class"
+        pass
+
+    @staticmethod
+    def _parse_owner(item):
+        owner = item.get('owner') or {}
+        if owner:
+            return {
+                'uploader': owner.get('name'),
+                'uploader_id': owner.get('id'),
+            }
+        return {}
+
+    def _fetch_page(self, list_id, page):
+        page += 1
+        items = self._call_api(list_id, 'page %d' % page, {
+            'page': page,
+            'pageSize': self._PAGE_SIZE,
+        })['items']
+        for video in items:
+            # this is needed to support both mylist and user
+            video = traverse_obj(video, ('video',), (), expected_type=dict)
```

Comment:
can't it be rewritten with one `traverse_obj` invocation?

Reply:
try `traverse_obj(items, (..., ('video', None)), expected_type=dict)`

Reply:
The last path should behave "get `video` if there is, if not, return the object itself", but yours isn't
In brief python code, it looks like
```python
items = ...['items']
for item in items:
  video = item.get('video') or item
  ...
```

Reply:
You are right. it cant be done without a for
```
>>> from yt_dlp.utils import traverse_obj
>>> items = [{'video': 'x'}, 'y']
>>> traverse_obj(items, (..., ('video', None)))
['x', {'video': 'x'}, 'y']
>>> [traverse_obj(i, 'video', None) for i in items]
['x', 'y']
```

Reply:
but there's no problem since it checks for `id` for each items